src/main/java/dev/morling/onebrc/CalculateAverage_merykittyunsafe.java

```diff
@@ -0,0 +1,379 @@
+            if (nodeSize == 0) {
+                aggrMap.insertInto(entryOffset, address, keySize);
+                break;
+            }
+
+            if (nodeSize != keySize) {
+                continue;
+            }
+
+            var nodeKey = ByteVector.fromArray(BYTE_SPECIES, aggrMap.data, (int) (entryOffset + PoorManMap.KEY_OFFSET));
+            long eqMask = line.compare(VectorOperators.EQ, nodeKey).toLong();
+            long validMask = semicolons ^ (semicolons - 1);
+            if ((eqMask & validMask) == validMask) {
+                break;
+            }
+        }
+
+        return parseDataPoint(aggrMap, entryOffset, address + keySize + 1);
+    }
+
+    // Process all lines that start in [offset, limit)
+    private static PoorManMap processFile(MemorySegment data, long offset, long limit) {
+        var aggrMap = new PoorManMap();
+        long base = data.address();
+        long begin = base + offset;
+        long end = base + limit;
+        // Find the start of a new line
+        if (offset != 0) {
+            begin--;
+            while (begin < end) {
+                if (UNSAFE.getByte(begin++) == '\n') {
+                    break;
+                }
+            }
+        }
+
+        // If there is no line starting in this segment, just return
+        if (begin == end) {
+            return aggrMap;
+        }
+
+        // The main loop, optimized for speed
+        while (begin < end - Math.max(BYTE_SPECIES.vectorByteSize(),
+                Long.BYTES + 1 + KEY_MAX_SIZE)) {
+            begin = iterate(aggrMap, begin);
+        }
+
+        // Now we are at the tail, just be simple
+        while (begin < end) {
+            int keySize = 0;
+            while (UNSAFE.getByte(begin + keySize) != ';') {
+                keySize++;
+            }
+            long entryOffset = aggrMap.indexSimple(begin, keySize);
+            begin = parseDataPointSimple(aggrMap, entryOffset, begin + 1 + keySize);
+        }
+
+        return aggrMap;
+    }
+
+    public static void main(String[] args) throws InterruptedException, IOException {
+        int processorCnt = Runtime.getRuntime().availableProcessors();
+        var res = new TreeMap<String, Aggregator>();
+        try (var file = FileChannel.open(Path.of(FILE), StandardOpenOption.READ);
+                var arena = Arena.ofShared()) {
+            var data = file.map(MapMode.READ_ONLY, 0, file.size(), arena);
+            long chunkSize = Math.ceilDiv(data.byteSize(), processorCnt);
+            var threadList = new Thread[processorCnt];
+            var resultList = new PoorManMap[processorCnt];
+            for (int i = 0; i < processorCnt; i++) {
+                int index = i;
+                long offset = i * chunkSize;
+                long limit = Math.min((i + 1) * chunkSize, data.byteSize());
+                var thread = new Thread(() -> resultList[index] = processFile(data, offset, limit));
```

Comment:
Noob here: I've heard in the past that lambda initialization cost is significant compared to anonymous classes. Is that not an issue here?

Reply:
lambda bootstrapping takes a short while once, but since invokedynamic is involved, each subsequent lambda "creation" takes the short path. anonymous inner classes are probably slightly faster, but it won't really matter here since that's just a few nanos at the start of the program. if you're doing this kind of optimization, you're using the wrong language

Reply:
I don't agree on the language choice. Read the assembly, and if there is a way to achieve it, just do it. Languages are just tools and the same for compilers; it's different in case the effort to achieve the effect is too much, but in this case, is a one liner.